### ISSUE TITLE
Roadmap RFC

### ISSUE THREAD

**Issue by dcrosta** [edelight/chef-mongodb#41]
Over the weekend, I created a 1.0.0 milestone, and tagged the existing issues and pull requests that I feel should be included in the 1.0 release. Of course this is up for some interpretation, but I wanted to let everyone know in case GitHub didn't notify appropriately (and if it spammed excessively as a result, my apologies).

There's one issue absent from the roadmap which I want to bring up here: the limitation of one `mongodb_instance` per node. This isn't actually enforced anywhere (so far as I can see), but it is implied by the way the recipes operate. In many cases, it may make sense to run several instances on the same node: multiple arbiters, config servers sharing space with shard servers, multiple replica set members (presumably of different replica sets), etc.

So, my first questions to everyone are: do you think that this is something this cookbook ought to support? Are there any configurations we should explicitly disallow?

I believe the major challenge in achieving this is reviewing where and how node attributes are used. Some attributes, such as the directory attributes `log_dir`, `init_dir`, `defaults_dir` are clearly shareable between instances (with appropriate namespacing of files within those dirs, of course); others set policies that may be considered global across a mongodb deployment: `enable_rest`, `cluster_name`, and `key_file` (if/when the appropriate pull request is merged); still others will likely be specific to a given instance: `port`, `dbpath`, and (hypothetically) `replica_set_name`.

One possible approach to alleviating these issues is to force values out of node attributes and into resource invocations. If we choose to go this route (and here's where it ties back into the Roadmap), I suggest we move more quickly towards migrating the code to provide LWRPs rather than a single definition. Though the LWRP(s?) would do what the definition currently does, taking the time to think about how we want users to use the cookbook now, before the 1.0.0 release, will be beneficial, and will help us prevent a surprising post-1.0 change to the interface.

So I guess my second question is: does anyone agree?

**Comment by jamesonjlee** [comment]
If I am interpreting this correctly, I am for moving settings that goes beyond installing a mongodb instance into LWRP(s).  I think it would help to clarify that this cookbook provides the installation, and configurations beyond installation should be done in a separate cookbook that is customized for your setup.

of course, this means that a sample cookbook on using the mongo cookbook to cluster/shard/replicate will have to be included somewhere (or in the readme).

**Comment by dcrosta** [comment]
Yes, I realize my first post may have been tl;dr -- so here's an executive summary:
1. We should support the ability to manage multiple mongodb things on a single node (two arbiters, arbitrer + config, etc). This is something people want to do (I claim).
2. Use node attributes only for things global to a node (package name, default dirs, etc). Push all other configurations down into the definition or LWRP (if we choose to go that route now).
3. Provide clear and concise directions or a sample cookbook showing how to use the new cookbook resources (good point @jamesonjlee)

**Comment by jamesonjlee** [comment]
The cookbook is going stale, so let's get stuff moving along with updates:

### 0.12.0 (immediate)
- Organize pull requests
- Merge in working bug fixes

### Tasks for 1.0.0 release
1. add test-kitchen support (via minitests)
2. LWRP the cookbook
3. check compatibility for ubuntu 12.04, centos 6
4. update documentation for LWRP
5. create chef-mongodb-sample to show how to use the mongodb recipe to roll your own mongodb

**Comment by jamesonjlee** [comment]
For fixing Debian(Ubuntu) and RedHat(CentOS) problems and streamlining the deployments:
1. Stop dropping in our own init systems
2. Utilizing /etc/sysconfig/mongodb and /etc/default/mongodb to load additional **system variables**
3. Use /etc/mongodb.conf to load **configuration variables** (because 10gen and repository sources from these)

I am going to create a 1.0.0-rc branch to implement this, and hopefully we can get more advanced tests in as well.

On top after will be LWRP of cookbook, documentation, and sample cookbook.

**Comment by josephholsten** [comment]
+1 for consistancy with upstream init and config vars.

If you see anything that needs testing but don't have time for it, feel free to mention me in the ticket. I especially want to get chef-solo-search integrated for minimal replicaset testing.

It might be good to do a full once-over of the internal features and ensure we're properly documenting them. The fact that a minimal default test case on ubuntu required the `10gen_repo` recipe makes me feel like either the docs or the defaults are lacking.

**Comment by jamesonjlee** [comment]
so After reading through the mongo docs, and the going through the all the recipes usage and definitions, we **have to** maintain our own init.d files.  The init files that comes with mongodb only runs mongod and not mongos, because [10gen does not provide a default mongos init files](http://docs.mongodb.org/manual/tutorial/install-mongodb-on-red-hat-centos-or-fedora-linux/#control-mongos) I thought this was fixed, but I guess I misunderstood what I heard at the last MongoNYC and MongoSF.

I still believe though we will follow sysconfig and dbconfig being **outside** of the init file, to make it easier to track changes from upstream init files.

I think any real deployment of mongo will use 10gen_repo (also, CentOS doesn't have mongodb in their default repository).  Somewhere down the line after the 1.0.0 we should change the default installer from default package to 10gen package (After testing that those changes doesn't break standard tests ofcourse).

**Comment by josephholsten** [comment]
Fully agree about the init script and repo, but we should def support folks who manage their repo mirror.

**Comment by jamesonjlee** [comment]
[Taken from @josephholsten suggestion in some remote PR somewhere]

As for ?? opinion on all the open PRs:

## progress
- #12 is deprecated in favor of #64 [closed]
- #34 is deprecated in favor of this refactor [closed]
- #43 is reasonable (if a hack) [merged]
- #47 is questionable (I mostly am skeptical of the port math) [closed in favor of #108]
- #64 is reasonable (preferred to #12) [merged](woo mms support... finally?)
- #77 is reasonable [merged]
- #80 is reasonable (requires merge) [merged](i find myself... screwing up these merges continuously)
- #83 is reasonable [closed](keeping the cookbook vanilla as possible from providers)
- #91 is reasonable (though half of it is logging changes) [merged]
- #111 is deprecated in favor of #91 [closed]
- #112 is terrible ;-) [closed]
- #119 is reasonable [merged](the sort would need to be removed or considered when search is moved out)
- #120 is reasonable (requires whitespace changes) [merged](white space fixing soon...)
- #131 is questionable (seems like a better solution exists) [merged](mongos needs a more looking over)
- #133 minor [merged]
- #138 is reasonable [merged]
- #142 is reasonable [merged](this needs a followup to make the hook url available or something)
- #144 is reasonable [merged]

## up for debate
- #65 is reasonable (mine)
- #82 is reasonable (may require a full rewrite)
- #93 needs rework, but worth ripping from
- #96 needs rework, but worth ripping from
- #102 needs rework (should use platform_family)
- #108 is reasonable (mostly whitespace)
- #125 needs rework (much of it was resolved in other patches)
- #136 needs rework (whitespace, unnecessary sh -> bash, &c)

**Comment by josephholsten** [comment]
I handled #82 with #192.

#102 should be simple to rebase

#108 is much more sane if viewed here: https://github.com/edelight/chef-mongodb/pull/108/files?w=0

**Comment by josephholsten** [comment]
# 2014-01-15 Update

## Pull Requests

### Ready to Merge
- #65 add mongodb/reload_action attribute to enable config changes without a restart
- #195 Add chefspec
- #203 The platform introspection mixin was added in Chef 11 for attributes files
- #204 upstart init files are 644
- #205 use the default service name for shard and config server
- #206 Variablize templates
- #207 Sanitize replicaset name (incorporates #171 make sure the replica set name makes it to the config file)
- #208

### Work Required
- #93 needs rework
- #96 needs rework
- #125 needs rework
- #194 needs rework, will be broken into pieces

## Issues
- #2 should be investigated (feature)
- #6 should be implemented and scheduled for inclusion in next major (improvement,breaking)
- #13 should be investigated (feature)
- #16 needs re-verification, it may also be handled by #205 (bug)
- #25 is being addressed by #194, and should be scheduled for inclusion in next major (feature,breaking)
- #41 is this document (doc)
- #52 should be investigated, fixed & tested (see #208) (bug)
- #54 should be investigated (feature)
- #95 should be implemented and scheduled for inclusion in next major (improvement,breaking)
- #117 needs acceptance criteria and to be broken into pieces (test)
- #118 needs verification (bug)
- #122 should be investigated and implemented (test)
- #124 should be investigated, fixed & tested (bug)
- #127 should be investigated and implemented (feature)
- #141 should be investigated and implemented (feature)
- #145 is waiting for response (feature)
- #152 is waiting on search abstraction (feature)
- #153 needs acceptance criteria and to be implemented (improvement)
- #158 is waiting for response (feature)
- #165 should be investigated (feature)
- #166 should be investigated and combined into #194 (feature)
- #169 should be fixed by #207 (bug)
- #173 should be investigated, fixed & tested (bug)
- #174 should be fixed by #201 & 205 (bug)
- #176 should be investigated, fixed & tested (bug)
- #177 makes me sad
- #179 should be implemented and scheduled for inclusion in next major (improvement, breaking)
- #181 needs replication steps (bug)
- #182 needs verification (bug)
- #186 should be investigated and implemented (feature)
- #189 needs verification (bug)
- #196 should be investigated & verified (bug)
- #199 should be investigated, fixed & tested (bug)

## Broad Efforts

most of these tickets are related to large, ongoing improvement efforts:
- Converting to a first-class chef resource, which can support multiple processes on the same box
- Abstracting search/supporting chef-solr
- Improving replicaset or shard configuration
- Improving testing
- Documentation

**Comment by josephholsten** [comment]
@jamesonjlee can you give me the ability to label issues? It would be a lot easier to keep this roadmap up to date if I could tag things.

**Comment by jamesonjlee** [comment]
@josephholsten I wish I could @edelight @thekorn would be the person who can, I only have contributor access.
